overlord/snapstate/autorefresh_gating.go

```diff
@@ -32,6 +35,249 @@ import (
 
 var gateAutoRefreshHookName = "gate-auto-refresh"
 
+// cumulative hold time for snaps other than self
+const maxOtherHoldDuration = time.Hour * 48
+
+type holdInfo struct {
+	// FirstHeld keeps the time when the given snap was first held for refresh by a gating snap.
+	FirstHeld time.Time
+	// HoldUntil stores the desired end time for holding.
+	HoldUntil time.Time
+}
+
+var timeNow = func() time.Time {
+	return time.Now()
+}
+
+func lastRefreshed(st *state.State, snapName string) (*time.Time, error) {
+	var snapst SnapState
+	if err := Get(st, snapName, &snapst); err != nil {
+		return nil, fmt.Errorf("internal error, cannot get snap %q: %v", snapName, err)
+	}
+	if snapst.LastRefresh != nil {
+		return snapst.LastRefresh, nil
+	}
+	inf, err := snapst.CurrentInfo()
+	if err != nil {
+		return nil, err
+	}
+	fst, err := os.Stat(inf.MountFile())
+	if err != nil {
+		return nil, err
+	}
+	t := fst.ModTime()
+	return &t, nil
+}
+
+func refreshGating(st *state.State) (map[string]map[string]*holdInfo, error) {
+	// affecting snap -> gating snap(s) -> first-held/hold-until time
```

Comment:
we probably want to use more general terminology at the level of the map here:

// held snaps -> holding snaps -> ...